apps/web/src/pages/Users/UpdateUserPage/components/AddCommunityRoleDialog.tsx

```diff
@@ -0,0 +1,256 @@
+import { useState } from "react";
+import { FormProvider, useForm } from "react-hook-form";
+import { useIntl } from "react-intl";
+import PlusIcon from "@heroicons/react/24/outline/PlusIcon";
+import { useQuery } from "urql";
+
+import { Dialog, Button } from "@gc-digital-talent/ui";
+import { Combobox, Select } from "@gc-digital-talent/forms";
+import { notEmpty } from "@gc-digital-talent/helpers";
+import { toast } from "@gc-digital-talent/toast";
+import {
+  commonMessages,
+  errorMessages,
+  formMessages,
+  getLocalizedName,
+} from "@gc-digital-talent/i18n";
+import {
+  UpdateUserRolesInput,
+  UpdateUserRolesMutation,
+  Role,
+  User,
+  graphql,
+  RoleInput,
+} from "@gc-digital-talent/graphql";
+
+import { getFullNameHtml } from "~/utils/nameUtils";
+import adminMessages from "~/messages/adminMessages";
+
+import { UpdateUserDataAuthInfoType } from "../UpdateUserPage";
+import { isCommunityTeamable } from "./helpers";
+
+const AddCommunityRoleCommunities_Query = graphql(/* GraphQL */ `
+  query AddCommunityRoleCommunities {
+    communities {
+      id
+      name {
+        en
+        fr
+      }
+      teamIdForRoleAssignment
+    }
+  }
+`);
+
+interface FormValues {
+  roles: string[];
+  community: string | null;
+}
+
+interface AddCommunityRoleDialogProps {
+  user: Pick<User, "id" | "firstName" | "lastName">;
+  authInfo: UpdateUserDataAuthInfoType;
+  communityRoles: Role[];
+  onAddRoles: (
+    submitData: UpdateUserRolesInput,
+  ) => Promise<UpdateUserRolesMutation["updateUserRoles"]>;
+}
+
+const AddCommunityRoleDialog = ({
+  user,
+  authInfo,
+  communityRoles,
+  onAddRoles,
+}: AddCommunityRoleDialogProps) => {
+  const intl = useIntl();
+  const [isOpen, setIsOpen] = useState<boolean>(false);
+  const { id, firstName, lastName } = user;
+  const userName = getFullNameHtml(firstName, lastName, intl);
+
+  const methods = useForm<FormValues>({
+    defaultValues: {
+      roles: [],
+      community: null,
+    },
+  });
+
+  const {
+    handleSubmit,
+    watch,
+    formState: { isSubmitting },
+  } = methods;
+
+  const handleAddRoles = async (formValues: FormValues) => {
+    let { roles } = formValues;
+    // despite the typing, roles is just a string may be complicated by there being only one team based role available
+    if (!Array.isArray(roles)) {
+      roles = [roles];
+    }
+
+    const roleInputArray: RoleInput[] = roles.map((role) => {
+      return { roleId: role, teamId: formValues.community };
+    });
+
+    return onAddRoles({
+      userId: id,
+      roleAssignmentsInput: {
+        attach: roleInputArray,
+      },
+    }).then(() => {
+      setIsOpen(false);
+      toast.success(
+        intl.formatMessage({
+          defaultMessage: "Role(s) added successfully",
+          id: "/17wgm",
+          description:
+            "Message displayed to user when one or more roles have been added to a user",
```

Comment:
This looks like a message that can be extracted to be reused between the different dialogs

Reply:
Sounds reasonable to me, done in https://github.com/GCTC-NTGC/gc-digital-talent/pull/11584/commits/b5f505b0f5a16f356e68f0a56382576de8174e64